[dynamo] Mark autograd.Functions as inlineable despite skipfiles

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118801

**Goal**: For autograd.Functions defined in torch/, we want them to be inlineable (not graph break) even when they are not allowlisted in skipfiles. The specific target is ViewNestedFromBuffer, for NestedTensor.

**This PR**:
* Add `should_force_inline()` method and `force_inline` constructor argument to `BaseUserFunctionVariable`. This is used to store whether we should ignore the skipfiles rules for inlining and inline anyway - right now, just for autograd.Function forward and backward functions.
* Remove handling for `produce_trampoline_autograd_apply` because this is no longer used after #116897

**Justification for removing handling for produce_trampoline_autograd_apply**:
*Before*: handling for this function was how autograd.Functions could be made inline-able despite skipfiles. apply, forward, and backward were all wrapped using `produce_trampoline_autograd_[...]` functions, which were inlined. Code pointer below is how this used to work. https://github.com/pytorch/pytorch/blob/83e8a0721d83b67576235fa49dd78a787f63b525/torch/_dynamo/variables/misc.py#L366-L368
*Now*: After #116897, a code search finds `produce_trampoline_autograd_apply` only in two spots: variables/misc.py and symbolic_convert.py. In symbolic_convert.py, it's just used to check whether an origin, so let's ignore that use. In variables/misc.py, it is used to wrap the apply function and then it is put directly into the graph, i.e. not inlined: https://github.com/pytorch/pytorch/blob/e426924c19500011aa419accc251a923fe857b94/torch/_dynamo/variables/misc.py#L400-L410. We won't try to inline this code, so we don't need the check any more.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

davidberard98 committed:
[dynamo] Mark autograd.Functions as inlineable despite skipfiles

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Mark autograd.Functions as inlineable despite skipfiles"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Mark autograd.Functions as inlineable despite skipfiles"


**Goal**: For autograd.Functions defined in torch/, we want them to be inlineable (not graph break) even when they are not allowlisted in skipfiles. The specific target is ViewNestedFromBuffer, for NestedTensor.

**This PR**:
* Add `should_force_inline()` method and `force_inline` constructor argument to `BaseUserFunctionVariable`. This is used to store whether we should ignore the skipfiles rules for inlining and inline anyway - right now, just for autograd.Function forward and backward functions.
* Remove handling for `produce_trampoline_autograd_apply` because this is no longer used after #116897

**Justification for removing handling for produce_trampoline_autograd_apply**:
*Before*: handling for this function was how autograd.Functions could be made inline-able despite skipfiles. apply, forward, and backward were all wrapped using `produce_trampoline_autograd_[...]` functions, which were inlined. Code pointer below is how this used to work. https://github.com/pytorch/pytorch/blob/83e8a0721d83b67576235fa49dd78a787f63b525/torch/_dynamo/variables/misc.py#L366-L368
*Now*: After #116897, a code search finds `produce_trampoline_autograd_apply` only in two spots: variables/misc.py and symbolic_convert.py. In symbolic_convert.py, it's just used to check whether an origin, so let's ignore that use. In variables/misc.py, it is used to wrap the apply function and then it is put directly into the graph, i.e. not inlined: https://github.com/pytorch/pytorch/blob/e426924c19500011aa419accc251a923fe857b94/torch/_dynamo/variables/misc.py#L400-L410. We won't try to inline this code, so we don't need the check any more.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Mark autograd.Functions as inlineable despite skipfiles"


**Goal**: For autograd.Functions defined in torch/, we want them to be inlineable (not graph break) even when they are not allowlisted in skipfiles. The specific target is ViewNestedFromBuffer, for NestedTensor.

**This PR**:
* Add `should_force_inline()` method and `force_inline` constructor argument to `BaseUserFunctionVariable`. This is used to store whether we should ignore the skipfiles rules for inlining and inline anyway - right now, just for autograd.Function forward and backward functions.
* Remove handling for `produce_trampoline_autograd_apply` because this is no longer used after #116897

**Justification for removing handling for produce_trampoline_autograd_apply**:
*Before*: handling for this function was how autograd.Functions could be made inline-able despite skipfiles. apply, forward, and backward were all wrapped using `produce_trampoline_autograd_[...]` functions, which were inlined. Code pointer below is how this used to work. https://github.com/pytorch/pytorch/blob/83e8a0721d83b67576235fa49dd78a787f63b525/torch/_dynamo/variables/misc.py#L366-L368
*Now*: After #116897, a code search finds `produce_trampoline_autograd_apply` only in two spots: variables/misc.py and symbolic_convert.py. In symbolic_convert.py, it's just used to check whether an origin, so let's ignore that use. In variables/misc.py, it is used to wrap the apply function and then it is put directly into the graph, i.e. not inlined: https://github.com/pytorch/pytorch/blob/e426924c19500011aa419accc251a923fe857b94/torch/_dynamo/variables/misc.py#L400-L410. We won't try to inline this code, so we don't need the check any more.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2261,19 +2261,13 @@ def inline_call(cls, parent, func, args, kwargs):
 
     @staticmethod
-    def check_inlineable(func):
+    def check_inlineable(func) -> skipfiles.SkipResult:
```
This approach looks good to me, but I have a pending work to merge allow_in_graph/inline/skip check into the unified ```trace_rules.lookup```. I'll move all of the ```check_inlineable``` logics into that function. Then we'll wrap a python function as ```UserFunctionVariable``` only if ```trace_rules.lookup``` tells us to inline it, and wrap as ```SkipFilesVariable``` if ```trace_rules.lookup``` tells us to skip it. I think it would help us make code more concise. I had the draft PR locally, but there are several trivial issues need to fix. Feel free to merge this PR if it blocks your work, I can rebase when I submit my PR.

davidberard98 replied:
How long do you think until your PR will be ready?

yanboliang replied:
Next week

davidberard98 replied:
I can probably wait then

davidberard98 committed:
Update on "[dynamo] Mark autograd.Functions as inlineable despite skipfiles"


**Goal**: For autograd.Functions defined in torch/, we want them to be inlineable (not graph break) even when they are not allowlisted in skipfiles. The specific target is ViewNestedFromBuffer, for NestedTensor.

**This PR**:
* Add `should_force_inline()` method and `force_inline` constructor argument to `BaseUserFunctionVariable`. This is used to store whether we should ignore the skipfiles rules for inlining and inline anyway - right now, just for autograd.Function forward and backward functions.
* Remove handling for `produce_trampoline_autograd_apply` because this is no longer used after #116897

**Justification for removing handling for produce_trampoline_autograd_apply**:
*Before*: handling for this function was how autograd.Functions could be made inline-able despite skipfiles. apply, forward, and backward were all wrapped using `produce_trampoline_autograd_[...]` functions, which were inlined. Code pointer below is how this used to work. https://github.com/pytorch/pytorch/blob/83e8a0721d83b67576235fa49dd78a787f63b525/torch/_dynamo/variables/misc.py#L366-L368
*Now*: After #116897, a code search finds `produce_trampoline_autograd_apply` only in two spots: variables/misc.py and symbolic_convert.py. In symbolic_convert.py, it's just used to check whether an origin, so let's ignore that use. In variables/misc.py, it is used to wrap the apply function and then it is put directly into the graph, i.e. not inlined: https://github.com/pytorch/pytorch/blob/e426924c19500011aa419accc251a923fe857b94/torch/_dynamo/variables/misc.py#L400-L410. We won't try to inline this code, so we don't need the check any more.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

yanboliang commented:
@davidberard98 Is that possible you can add something like:
```
if isinstance(obj. torch.autograd.Function) and obj.__code__.co_name == "apply":
    return UserFunctionVariable
```
in ```trace_rules.lookup_inner()``` to address this issue? I think this would be more concise.

davidberard98 commented:
@yanboliang I'm not actually sure how to tell that a MyAutogradFunction.forward is part of MyAutogradFunction; it doesn't have a `__self__`, and `__class__` doesn't return anything useful.

davidberard98 committed:
Update on "[dynamo] Mark autograd.Functions as inlineable despite skipfiles"


**Goal**: For autograd.Functions defined in torch/, we want them to be inlineable (not graph break) even when they are not allowlisted in skipfiles. The specific target is ViewNestedFromBuffer, for NestedTensor.

**This PR**:
* Add `should_force_inline()` method and `force_inline` constructor argument to `BaseUserFunctionVariable`. This is used to store whether we should ignore the skipfiles rules for inlining and inline anyway - right now, just for autograd.Function forward and backward functions.
* Remove handling for `produce_trampoline_autograd_apply` because this is no longer used after #116897

**Justification for removing handling for produce_trampoline_autograd_apply**:
*Before*: handling for this function was how autograd.Functions could be made inline-able despite skipfiles. apply, forward, and backward were all wrapped using `produce_trampoline_autograd_[...]` functions, which were inlined. Code pointer below is how this used to work. https://github.com/pytorch/pytorch/blob/83e8a0721d83b67576235fa49dd78a787f63b525/torch/_dynamo/variables/misc.py#L366-L368
*Now*: After #116897, a code search finds `produce_trampoline_autograd_apply` only in two spots: variables/misc.py and symbolic_convert.py. In symbolic_convert.py, it's just used to check whether an origin, so let's ignore that use. In variables/misc.py, it is used to wrap the apply function and then it is put directly into the graph, i.e. not inlined: https://github.com/pytorch/pytorch/blob/e426924c19500011aa419accc251a923fe857b94/torch/_dynamo/variables/misc.py#L400-L410. We won't try to inline this code, so we don't need the check any more.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
